### DISCUSSION TITLE
How to handle commands and events from the third party

### DISCUSSION THREAD

**Discussion by nyh1030** [fraktalio/fmodel#259] [Q&A]
hello.
I'm having a hard time implementing 
`Aggregate` and `MaterializedView` should be set up in this case.

Before I explain the issue, here is my situation.

I have an Aggregate with only one `Order`.
`OrderDecider`, and `OrderView`.
`OrderDecider` is composed of `OrderCommand`, `OrderEvent`, and inherits and implements `Command` and `Event`.

I also added a Saga for external integration, which I'll call `PaymentSaga` for simplicity.
`PaymentCommand` and `PaymentEvent` also inherit from 
`Command` and `Event` inherit and implement them.

---
**My issue : Problem setting `Aggregate`, `MaterizlizedView` when there is only one Aggregate in `monolith` method.**
<img width="1257" alt="image" src="https://github.com/fraktalio/fmodel/assets/10717383/ff6f20f8-4d16-4c5b-af92-e1654095f066">

Details : 
The example `combines` Order and Restaurant to create a Payment. 
and processed as `Command` and `Event`,
In my situation, I only have Order, so I can't process it with `.combine()`.

In `OrderDecider`, `OrderView`.
`OrderCommand`, and `OrderEvent`, I can set them to handle `Command` and `Event` instead.
But then I need to handle `PaymentCommand` and `PamentEvent` as well.

In `EventStream` and `MaterializedViewStateRepositoryImpl`, I want to keep them to handle the entire Event.
without `.combine()` Aggregate, MaterizlizedView for setting up a Is there a good way?

I will summarize the question as mine is too long.
**1 Aggregate, 1 Third Party Saga** 
in the case of Command, Event type, and 
Aggregate, MaterializedView in the case of a Third Party Saga.

[Note]
I explicitly specified the co-variance of generic types in Kotlin, but it didn't solve the whole problem.
Example.) 
```
class EventSourcingLockingAggregate<Command, Aggregate, out Event, Id>(...) { ... }
```

And additionally
I was wondering if I don't need to implement Decider and View for `Payment(the third party)` separately.

**Comment by idugalic** [comment]
Hi @nyh1030 

Looks like your OrderDecider and OrderSaga generic types don't match. 

1. Your EventSourcingLockingOrchestratingAggregate is needed only in case of `combined` decider/decider that is combined out of two, for example, `OrderDecider combine RestaurantDecider`. Only in that case Saga is needed, to orchestrate and execute these deciders, and produce all events in one transaction. These events will be stored then, in one transaction. This aggregate is not communicating with external systems in any way. 
2. Beter, use `EventSourcingLockingAggregate<OrderCommand, Order, OrderEvent>` (not orchestrating/no saga as parameter) for the OrderDecider only.  And, create an independent SagaManager for` Saga<OrderEvent, PaymentCommand>` . Check the example: [https://github.com/fraktalio/fmodel-ktor-demo/blob/main/src/main/kotlin/com/fraktalio/application/SagaManager.kt](https://github.com/fraktalio/fmodel-ktor-demo/blob/main/src/main/kotlin/com/fraktalio/application/SagaManager.kt)
3. Now, you have OrderAggreggate and PaymentSagaManager as two components. This is good in my opinion, as PayamentSaga is asynchronously integrated with third-party service, minimizing the runtime coupling of the two systems. Saga acts as Anti-corruption layer as well. 

Additionally, I don't see a need for a dedicated Payment Decider/View. It could be good enough that the concept of Payment is modeled/implemented as a simple Enum/Status attribute in your OrderDecider. In that case, OrderDecider should have two additional commands: MarkOrderAsPaid, and MarkOrderAsNotPaid, for example. 

Best,
Ivan
